File: 113.path-sum-ii.js
/*
 * @lc app=leetcode id=113 lang=javascript
 *
 * [113] Path Sum II
 */

function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

// @lc code=start
/**
 * @param {TreeNode} root
 * @param {number} sum
 * @return {number[][]}
 */
var pathSum = function (root, sum) {
  const helper = (cur, node, target) => {
    if (!node) {
      return;
    }
    if (!node.left && !node.right && node.val === target) {
      res.push([...cur, node.val]);
    }
    cur.push(node.val);
    helper(cur, node.left, target - node.val);
    helper(cur, node.right, target - node.val);
    cur.pop();
  };
  const res = [];
  helper([], root, sum);
  return res;
};
// @lc code=end

File: 113.path-sum-ii.py
#
# @lc app=leetcode id=113 lang=python3
#
# [113] Path Sum II
#

# Time:  O(n)
# Space: O(h)

from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


# @lc code=start
class Solution:
    def pathSum(self, root: TreeNode, targetSum: int) -> List[List[int]]:
        def helper(cur: list, node: TreeNode, target):
            if not node:
                return
            if not node.left and not node.right and node.val == target:
                res.append(cur + [node.val])
            cur.append(node.val)
            helper(cur, node.left, target - node.val)
            helper(cur, node.right, target - node.val)
            cur.pop()

        res = []
        helper([], root, targetSum)
        return res


# @lc code=end
